### PR TITLE
Candy Raffle

### PR DESCRIPTION
**Pull request by murderteeth**
## Description
Add UIs for candy raffle, blood sacrifice, and skin drop results

## Type of change
- [x] New feature (non-breaking change which adds functionality)

## Change details
Replace the Spooky Festival tile with a link to the Candy Raffle. Add screens for raffle, sacrifice, and skin drop results. Adds a temp address for Raffle.sol here https://github.com/murderteeth/RarityExtended/blob/feature/skinDrop/next.config.js#L73

## Resources
Here are the scripts I used to deploy Raffle.sol and simulate results on localhost:
https://github.com/murderteeth/rarity_skins_drop/tree/mt/scripts/localhost

## ⚠ Not quite working yet
Raffle.winningOdds() was reverting (with no reason string) when not enough sums/wallets are entered in the lottery. After entering a bunch, it seems to work. I skipped it for now, it's commented here: https://github.com/murderteeth/RarityExtended/blob/feature/skinDrop/pages/festivals/raffle/index.js#L59

After a sacrifice we'd like to remove the sacrificed summoner from the UI. I tried a few things and landed on this:
https://github.com/murderteeth/RarityExtended/blob/feature/skinDrop/pages/festivals/raffle/sacrifice.js#L37
That still doesn't work tho.

### CODE REVIEW

File: components/ButtonCounter.js [hunk 1]
```diff
@@ -6,6 +6,7 @@
 ******************************************************************************/
 
 import	React		from	'react';
+import ButtonCounterBase	 from	 './ButtonCounterBase';
```
Comment (Major-Eth) [inline components/ButtonCounter.js:9]: Could you use absolute import here, instead of relative? 🙏

File: components/ButtonCounter.js [hunk 2]
```diff
@@ -34,105 +34,14 @@ function	ButtonCounter({
...
-		</div>
-	);
+	return <ButtonCounterBase 
+		className={className}
+		threshold={threshold}
+		value={value}
+		onIncrement={onIncrease}
+		onDecrement={onDecrease}
+		backgroundColor={backgroundColor}
+		></ButtonCounterBase>;
```
Comment (Major-Eth) [inline components/ButtonCounter.js:44]:
```suggestion
		/>;
```

File: components/ModalSelectAdventurer.js [hunk 3]
```diff
@@ -13,6 +6,7 @@ import	useRarity						from	'contexts/useRarity';
 import	useWeb3							from	'contexts/useWeb3';
 import	useLocalStorage					from	'hook/useLocalStorage';
 import	CLASSES							from	'utils/codex/classes';
+import { useEffect } from 'react/cjs/react.development';
```
Comment (Major-Eth) [inline components/ModalSelectAdventurer.js:9]:
```suggestion

```

File: components/ModalSelectAdventurer.js [hunk 4]
```diff
@@ -1,10 +1,3 @@
-/******************************************************************************
-**	@Author:				Rarity Extended
-**	@Twitter:				@RXtended
-**	@Date:					Thursday September 23rd 2021
-**	@Filename:				ModalCurrentAdventurer.js
-******************************************************************************/
-
 import	React, {Fragment, useState}		from	'react';
```
Comment (Major-Eth) [inline components/ModalSelectAdventurer.js:1]:
```suggestion
import	React, {Fragment, useState, useEffect}		from	'react';
```

File: utils/actions/candyRaffle.js [hunk 5]
```diff
@@ -0,0 +1,212 @@
+import { ethers } from 'ethers'
+import toast from 'react-hot-toast'
+
+const CONFIRMATIONS = 1
```
Comment (Major-Eth) [inline utils/actions/candyRaffle.js:4]:
```suggestion
const CONFIRMATIONS = 2
```
Prevent the caching from the free RPC to screw up things

**Review by Major-Eth — CHANGES_REQUESTED** [review]

File: components/ModalSelectAdventurer.js [hunk 6]
```diff
@@ -125,6 +124,9 @@ function ModalCurrentAdventurer({isOpen, closeModal}) {
 								
 							<div className={'grid grid-cols-2 md:grid-cols-4 gap-4 md:gap-8 gap-y-0 md:gap-y-4 min-h-0 md:min-h-133 max-h-72 md:max-h-133 overflow-y-scroll px-1'}>
 								{[...Object.values(rarities)]
+                  .filter((adventurer) => {
+                    return !exclusions.includes(adventurer.tokenID);
```
Comment (Major-Eth) [inline components/ModalSelectAdventurer.js:128]:
```suggestion
                    return !exclusions.includes(adventurer?.tokenID);
```

**Review by Major-Eth — CHANGES_REQUESTED** [review]

File: pages/festivals/raffle/sacrifice.js [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+      summonerToSacrificeName: currentAdventurer.name || currentAdventurer.tokenID,
+      summonerToReceive: beneficiary.tokenID,
+      summonerToReceiveName: beneficiary.name || beneficiary.tokenID
+    }, async () => {
+      beneficiary.inventory[9] = String(Number(beneficiary.inventory[9]) + CANDIES_PER_SUMMONER)
+      set_currentAdventurer(beneficiary)
+
+      // what's the best way to _remove_ a summoner from the rarities list?
+      // this doesn't seem to work..
+      delete rarities[currentAdventurer.tokenID]
```
Comment (Major-Eth) [inline pages/festivals/raffle/sacrifice.js:39]:
```suggestion

```

File: pages/festivals/raffle/sacrifice.js [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+      set_rarities({...rarities})
```
Comment (Major-Eth) [inline pages/festivals/raffle/sacrifice.js:40]:
```suggestion
			set_rarities((prev) => {
				delete prev[currentAdventurer.tokenID];
				return ({...prev});
			});
```

**Review by Major-Eth — CHANGES_REQUESTED** [review]
You should add this
```
if (toAddress(owner) !== toAddress(address)) {
    set_rarities((prev) => {
        delete prev[tokenID];
        return ({...prev})
    });
    return;
}
return
```

In `contexts/useRarity.js` L.223 (instead of just return) to remove the adventurer from previous mapping if no longer owner

File: pages/festivals/raffle/index.js [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+  function sellCandies() {
+    return <>
+      <div className={'w-96 flex flex-row items-center justify-between'}>
+        <div>{`Candies`}</div>
+        <div className={'flex flex-row items-center text-2xl'}>{candies} <CandyIcon className={'ml-4'} width={48} height={48} /></div>
+      </div>
+      <div className={'my-4 w-96 flex flex-row items-center justify-between'}>
+        <div>{`Odds`}</div>
+        <div className={'flex flex-row items-center'}>
+          {1}/{Math.round(odds[1] / odds[0])}
```
Comment (Major-Eth) [inline pages/festivals/raffle/index.js:106]:
```suggestion
          {1}/{odds[0] === 0 ? 1 : Math.round(odds[1] / odds[0])}
```

**Review by Major-Eth — CHANGES_REQUESTED** [review]

File: components/ButtonCounter.js (hunk 1 above)
Comment (murderteeth) [reply components/ButtonCounter.js]: 75136d6

File: components/ButtonCounter.js (hunk 2 above)
Comment (murderteeth) [reply components/ButtonCounter.js]: 713a24a

File: utils/actions/candyRaffle.js (hunk 5 above)
Comment (murderteeth) [reply utils/actions/candyRaffle.js]: 713a24a

File: components/ModalSelectAdventurer.js (hunk 6 above)
Comment (murderteeth) [reply components/ModalSelectAdventurer.js]: 713a24a

File: pages/festivals/raffle/index.js (hunk 9 above)
Comment (murderteeth) [reply pages/festivals/raffle/index.js]: 713a24a

File: pages/festivals/raffle/sacrifice.js (hunk 7 above)
Comment (murderteeth) [reply pages/festivals/raffle/sacrifice.js]: 77740c2

File: pages/festivals/raffle/sacrifice.js (hunk 8 above)
Comment (murderteeth) [reply pages/festivals/raffle/sacrifice.js]: 77740c2

**Comment by murderteeth** [comment]
@Major-Eth checkout 710c42a pls, changed the way odds are computed and shown

File: components/ModalSelectAdventurer.js (hunk 4 above)
Comment (Major-Eth) [reply components/ModalSelectAdventurer.js:1]: @murderteeth I need this one to be fixed to approve :)

**Review by Major-Eth — APPROVED** [review]

**Pull request merged**